File: DSA/GreedyAlgo/Code/JobSequencing.py
# Job Sequencing Problem in Python

class Job:
    def __init__(self, id, deadline, profit):
        self.id = id
        self.deadline = deadline
        self.profit = profit

def job_sequencing(jobs):
    # Sort by profit (descending)
    jobs.sort(key=lambda x: x.profit, reverse=True)
    
    n = len(jobs)
    slot = [-1] * n
    
    total_profit = 0
    jobs_selected = 0
    
    for job in jobs:
        for j in range(job.deadline - 1, -1, -1):
            if j < n and slot[j] == -1:
                slot[j] = job.id
                total_profit += job.profit
                jobs_selected += 1
                break
    
    print("Selected Jobs (in execution order):")
    for i in range(n):
        if slot[i] != -1:
            print(f"Slot {i + 1}: Job {slot[i]}")
    
    print(f"\nTotal Jobs: {jobs_selected}")
    print(f"Total Profit: {total_profit}")

# Test
print("=== Job Sequencing Problem (Python) ===\n")

jobs = [
    Job(1, 4, 20),
    Job(2, 1, 10),
    Job(3, 1, 40),
    Job(4, 1, 30),
    Job(5, 3, 35)
]

print("Jobs (ID, Deadline, Profit):")
for job in jobs:
    print(f"Job {job.id}: Deadline {job.deadline}, Profit {job.profit}")
print()

job_sequencing(jobs)

print("\n=== Greedy Strategy ===")
print("1. Sort jobs by profit (descending)")
print("2. For each job:")
print("   - Find latest available slot before deadline")
print("   - If slot found: schedule job there")
print("3. Calculate total profit\n")

print("=== Complexity ===")
print("Time Complexity:  O(n²) worst case")
print("Space Complexity: O(n)")
print("Optimal: Yes (greedy optimal for this problem)")

File: DSA/GreedyAlgo/Code/JobSequencing.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Job {
    int id;
    int deadline;
    int profit;
} Job;

int compare(const void* a, const void* b) {
    return ((Job*)b)->profit - ((Job*)a)->profit;
}

void jobSequencing(Job* jobs, int n) {
    qsort(jobs, n, sizeof(Job), compare);
    
    int* slot = (int*)malloc(n * sizeof(int));
    int* result = (int*)malloc(n * sizeof(int));
    
    for (int i = 0; i < n; i++) {
        slot[i] = 0;
        result[i] = -1;
    }
    
    int totalProfit = 0;
    int jobsSelected = 0;
    
    for (int i = 0; i < n; i++) {
        for (int j = jobs[i].deadline - 1; j >= 0; j--) {
            if (j < n && !slot[j]) {
                result[j] = jobs[i].id;
                slot[j] = 1;
                totalProfit += jobs[i].profit;
                jobsSelected++;
                break;
            }
        }
    }
    
    printf("Selected Jobs (in execution order):\n");
    for (int i = 0; i < n; i++) {
        if (result[i] != -1) {
            printf("Slot %d: Job %d\n", i + 1, result[i]);
        }
    }
    
    printf("\nTotal Jobs: %d\n", jobsSelected);
    printf("Total Profit: %d\n", totalProfit);
    
    free(slot);
    free(result);
}

int main() {
    printf("=== Job Sequencing Problem (C) ===\n\n");
    
    Job jobs[] = {
        {1, 4, 20},
        {2, 1, 10},
        {3, 1, 40},
        {4, 1, 30},
        {5, 3, 35}
    };
    int n = 5;
    
    printf("Jobs (ID, Deadline, Profit):\n");
    for (int i = 0; i < n; i++) {
        printf("Job %d: Deadline %d, Profit %d\n", jobs[i].id, jobs[i].deadline, jobs[i].profit);
    }
    printf("\n");
    
    jobSequencing(jobs, n);
    
    printf("\n=== Greedy Strategy ===\n");
    printf("1. Sort jobs by profit (descending)\n");
    printf("2. For each job:\n");
    printf("   - Find latest available slot before deadline\n");
    printf("   - If slot found: schedule job there\n");
    printf("3. Calculate total profit\n\n");
    
    printf("=== Complexity ===\n");
    printf("Time Complexity:  O(n²) worst case\n");
    printf("Space Complexity: O(n)\n");
    printf("Optimal: Yes (greedy optimal for this problem)\n");
    
    return 0;
}
